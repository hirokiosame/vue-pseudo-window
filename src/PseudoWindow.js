export default {
	name: 'pseudo-window',

	props: {
		document: Boolean
	},

	render() {
		const defSlot = this.$slots.default;
		return defSlot && defSlot.length === 1 ? defSlot[0] : defSlot;
	},

	data() {
		return { handlers: [] };
	},

	mounted() {
		this.bindEventListeners();
	},
	destroyed() {
		this.unbindEventListeners();
	},

	methods: {
		bindEventListeners() {
			var $listeners = this.$listeners;
			for (var event in $listeners) {
				if (!$listeners.hasOwnProperty(event)) {
					continue;
				}
				const e = this.normalizeEvent(
					this.document ? window.document : window,
					event,
					$listeners[event]
				);
				e.target.addEventListener(e.name, e.handler, e.opts);
				this.handlers.push(e);
			}
		},

		unbindEventListeners() {
			while (this.handlers.length) {
				const e = this.handlers.shift();
				e.target.removeEventListener(e.name, e.handler, e.opts);
			}
		},

		normalizeEvent(target, name, handler) {
			const passive = name.charAt(0) === '&';
			name = passive ? name.slice(1) : name;
			const once = name.charAt(0) === '~'; // Prefixed last, checked first

			name = once ? name.slice(1) : name;
			const capture = name.charAt(0) === '!';
			name = capture ? name.slice(1) : name;
			return {
				target: target,
				name: name,
				handler: handler,
				opts: {
					once: once,
					capture: capture,
					passive: passive
				}
			};
		}
	}
};
