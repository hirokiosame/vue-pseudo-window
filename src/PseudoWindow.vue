<script>
export default {
  name: "pseudo-window",

  props: {
    document: Boolean
  },

  render() {
    const { default: defSlot } = this.$slots;
    return defSlot && defSlot.length === 1 ? defSlot[0] : defSlot;
  },

  data() {
    return {
      handlers: []
    };
  },

  mounted() {
    this.bindEventListeners();
  },

  destroyed() {
    this.unbindEventListeners();
  },

  methods: {
    bindEventListeners() {
      Object.keys(this.$listeners).forEach(event => {
        const e = this.normalizeEvent(
          this.document ? window.document : window,
          event,
          this.$listeners[event]
        );
        e.target.addEventListener(e.name, e.handler, e.opts);
        this.handlers.push(e);
      });
    },

    unbindEventListeners() {
      while (this.handlers.length) {
        const e = this.handlers.shift();
        e.target.removeEventListener(e.name, e.handler, e.opts);
      }
    },

    normalizeEvent(target, name, handler) {
      const passive = name.charAt(0) === "&";
      name = passive ? name.slice(1) : name;
      const once = name.charAt(0) === "~"; // Prefixed last, checked first
      name = once ? name.slice(1) : name;
      const capture = name.charAt(0) === "!";
      name = capture ? name.slice(1) : name;
      return {
        target,
        name,
        handler,
        opts: {
          once,
          capture,
          passive
        }
      };
    }
  }
};
</script>
