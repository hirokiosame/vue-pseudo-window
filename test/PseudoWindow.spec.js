import { mount } from '@vue/test-utils';
import PseudoWindow from '@/PseudoWindow';

describe('Window', () => {
	it('should not catch "click" event', () => {
		const myMockFn = jest.fn();
		const wrapper = mount(PseudoWindow, {
			attachToDocument: true,
		});

		global.window.dispatchEvent(new Event('click'));
		expect(myMockFn).not.toBeCalled();
		expect(wrapper.element).toMatchSnapshot();
	});

	it('should catch "resize" event', () => {
		const myMockFn = jest.fn();
		const wrapper = mount(PseudoWindow, {
			attachToDocument: true,
			listeners: {
				resize: myMockFn,
			},
		});

		global.window.dispatchEvent(new Event('resize'));
		expect(myMockFn).toBeCalled();
		expect(wrapper.element).toMatchSnapshot();
	});

	it('should cleanup event handler and not catch "resize" event', () => {
		const myMockFn = jest.fn();
		const wrapper = mount(PseudoWindow, {
			attachToDocument: true,
			listeners: {
				resize: myMockFn,
			},
		});

		wrapper.destroy();

		global.window.dispatchEvent(new Event('resize'));
		expect(myMockFn).not.toHaveBeenCalled();
		expect(wrapper.element).toMatchSnapshot();
	});

	it('should catch "resize" event once', () => {
		const myMockFn = jest.fn();
		const wrapper = mount(PseudoWindow, {
			attachToDocument: true,
			listeners: {
				'~resize': myMockFn,
			},
		});

		global.window.dispatchEvent(new Event('resize'));
		global.window.dispatchEvent(new Event('resize'));
		expect(myMockFn.mock.calls.length).toBe(1);
		expect(wrapper.element).toMatchSnapshot();
	});

	it('should not catch "resize" event', () => {
		const myMockFn = jest.fn();
		const wrapper = mount(PseudoWindow, {
			attachToDocument: true,
			listeners: {
				'~resize': myMockFn,
			},
		});

		wrapper.destroy();

		global.window.dispatchEvent(new Event('resize'));
		expect(myMockFn).not.toBeCalled();
		expect(wrapper.element).toMatchSnapshot();
	});
});

describe('Document', () => {
	it('should not catch "resize" on window', () => {
		const myMockFn = jest.fn();
		const wrapper = mount(PseudoWindow, {
			attachToDocument: true,
			propsData: {
				document: true,
			},
			listeners: {
				resize: myMockFn,
			},
		});

		global.window.dispatchEvent(new Event('resize'));

		expect(myMockFn).not.toHaveBeenCalled();
		expect(wrapper.element).toMatchSnapshot();
	});

	it('should catch "click" on document', () => {
		const myMockFn = jest.fn();
		const wrapper = mount(PseudoWindow, {
			attachToDocument: true,
			propsData: {
				document: true,
			},
			listeners: {
				click: myMockFn,
			},
		});

		global.window.document.dispatchEvent(new Event('click'));

		expect(myMockFn).toHaveBeenCalled();
		expect(wrapper.element).toMatchSnapshot();
	});
});
