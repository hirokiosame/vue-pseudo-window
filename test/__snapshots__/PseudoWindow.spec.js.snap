// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document should catch "click" on document 1`] = `<!---->`;

exports[`Document should not catch "resize" on window 1`] = `<!---->`;

exports[`Window should catch "resize" event 1`] = `<!---->`;

exports[`Window should catch "resize" event once 1`] = `<!---->`;

exports[`Window should not catch "resize" event 1`] = `<!---->`;
